BlockSettleUILib/BSTerminalMainWindow.cpp

```diff
@@ -2303,6 +2198,174 @@ void BSTerminalMainWindow::processDeferredDialogs()
+
+   connect(ui_->widgetChat, &ChatWidget::emailHashRequested, bsClient_.get(), &BsClient::findEmailHash);
+   connect(bsClient_.get(), &BsClient::emailHashReceived, ui_->widgetChat, &ChatWidget::onEmailHashReceived);
+
+   connect(bsClient_.get(), &BsClient::processPbMessage, orderListModel_.get(), &OrderListModel::onMessageFromPB);
+
+   utxoReservationMgr_->setFeeRatePb(result.feeRatePb);
+   connect(bsClient_.get(), &BsClient::feeRateReceived, this, [this] (float feeRate) {
+      utxoReservationMgr_->setFeeRatePb(feeRate);
+   });
+
+   authManager_->setCelerClient(celerConnection_);
+
+   setLoginButtonText(currentUserLogin_);
+   setWidgetsAuthorized(true);
+
+   // We don't use password here, BsProxy will manage authentication
+   SPDLOG_LOGGER_DEBUG(logMgr_->logger(), "got celer login: {}", result.celerLogin);
+   celerConnection_->LoginToServer(bsClient_.get(), result.celerLogin, email);
+
+   ui_->widgetWallets->setUsername(currentUserLogin_);
+   action_logout_->setVisible(false);
+   action_login_->setEnabled(false);
+
+   // Market data, charts and chat should be available for all Auth eID logins
+   mdProvider_->SubscribeToMD();
+
+   connect(bsClient_.get(), &BsClient::processPbMessage, ui_->widgetChat, &ChatWidget::onProcessOtcPbMessage);
+   connect(ui_->widgetChat, &ChatWidget::sendOtcPbMessage, bsClient_.get(), &BsClient::sendPbMessage);
+
+   connect(bsClient_.get(), &BsClient::ccGenAddrUpdated, this, [this](const BinaryData &ccGenAddrData) {
+      ccFileManager_->setCcAddressesSigned(ccGenAddrData);
+   });
+
+   accountEnabled_ = true;
+   onAccountTypeChanged(result.userType, result.enabled);
+   connect(bsClient_.get(), &BsClient::accountStateChanged, this, [this](bs::network::UserType userType, bool enabled) {
+      onAccountTypeChanged(userType, enabled);
+   });
+
+   connect(bsClient_.get(), &BsClient::tradingStatusChanged, this, [](bool tradingEnabled) {
+      NotificationCenter::notify(tradingEnabled ? bs::ui::NotifyType::TradingEnabledOnPB : bs::ui::NotifyType::TradingDisabledOnPB, {});
+   });
+}
+
+const std::string &BSTerminalMainWindow::loginApiKey() const
+{
+   return loginApiKey_;
+}
+
+void BSTerminalMainWindow::initApiKeyLogins()
+{
+   if (loginTimer_ || loginApiKey().empty() || !gotChatKeys_) {
+      return;
+   }
+   loginTimer_ = new QTimer(this);
+   connect(loginTimer_, &QTimer::timeout, this, [this] {
+      tryLoginUsingApiKey();
+   });
+   tryLoginUsingApiKey();
+}
+
+void BSTerminalMainWindow::tryLoginUsingApiKey()
+{
+   if (loginApiKey().empty() || autoLoginState_ != AutoLoginState::Idle) {
+      return;
+   }
+
+   auto logger = logMgr_->logger("proxy");
+   auto client = createClient();
+
+   autoLoginState_ = AutoLoginState::Connecting;
+   connect(client.get(), &BsClient::connected, this, [this, client, logger] {
+      client->authorize(loginApiKey());
+      connect(client.get(), &BsClient::authorizeDone, this, [this, client, logger](bool success, const std::string &email) {
```

Comment:
Connecting a signal after the action which can emit it (think tests), may be missed in some rare cases. So I think better to connect all signals before doing anything with `client`. Also, nested connecting looks rather exotic.

Reply:
Sure, fixed.